docs: Add new section, tutorials, to left nav, and a tutorial on routing (#36545)
angular/angular: aio/content/guide/router-tutorial.md

[Singleton Services](/guide/singleton-services#forroot-and-the-router).

## Update your component with `router-outlet`

At this point, you have defined two routes for your application. However, your application
still has both the `crisis-list` and `heroes-list` components hard-coded in your `app.component.html` template. For your routes to
work, you need to update your template to dynamically load a component based on the URL path.

To implement this functionality, you add the `router-outlet` directive to your template file.

1. From your code editor, open the `app.component.html` file.

1. Delete the following lines.

   <code-example header="src/app/app.component.html" path="router-tutorial/src/app/app.component.html" region="components"></code-example>

1. Add the `router-outlet` directive.

   <code-example header="src/app/app.component.html" path="router-tutorial/src/app/app.component.html" region="router-outlet"></code-example>

View your updated application in your browser. You should see only the application title. To
view the `crisis-list` component, add `crisis-list` to the end of the path in your browser's
address bar. For example:

> <code-example language="sh">

Reviewer: These are not shell commands. So, they should have no language attribute or have `language="none"`.

http://localhost:4200/crisis-list
</code-example>

Notice that the `crisis-list` component displays. Angular is using the route you defined to dynamically load the
component. You can load the `heroes-list` component the same way:

<code-example language="sh">
http://localhost:4200/heroes-list
</code-example>

## Control navigation with UI elements

Currently, your application supports two routes. However, the only way to use those routes
is for the user to manually type the path in the browser's address bar. In this section, you'll
add two links that users can click to navigate between the `heroes-list` and `crisis-list`
components. You'll also add some CSS styles. While these styles are not required, they make
it easier to identify the link for the currently-displayed component. You'll add that functionality
in the next section.

1. Open the `app.component.html` file and add the following HTML below the title.

   <code-example header="src/app/app.component.html" path="router-tutorial/src/app/app.component.html" region="nav"></code-example>

   This HTML uses an Angular directive, `routerLink`. This directive connects the routes
   you defined to your template files.

> 1. Open the `app.component.styles.css` file and add the following styles.

Reviewer: ```suggestion
1. Open the `app.component.css` file and add the following styles.
```

   * `redirectTo`. This property instructs Angular to redirect from an empty path to the
     `heroes-list` path.
   * `pathMatch`. This property instructs Angular on how much of the URL to match. For this
      tutorial, you should set this property to `full`. This strategy is recommended when
      you have an empty string for a path. For more information about this property,
      see the [Route API documentation](/api/router/Route).

Now when you open your application, it displays the `heroes-list` component by default.

## Adding a 404 page

It is possible for a user to try to access a route that you have not defined. To account for
this behavior, a best practice is to display a 404 page. In this section, you'll create a 404 page and
update your route configuration to show that page for any unspecified routes.

1. From the terminal, create a new component, `PageNotFound`.

   <code-example language="sh">
   ng generate component page-not-found
   </code-example>

1. From your code editor, open the `page-not-found.component.html` file and replace its contents
   with the following HTML.

>    <code-example header="src/app/pagenotfound/page-not-found.component.html" path="router-tutorial/src/app/page-not-found/page-not-found.component.html"></code-example>

Reviewer: ```suggestion
   <code-example header="src/app/page-not-found/page-not-found.component.html" path="router-tutorial/src/app/page-not-found/page-not-found.component.html"></code-example>
```

# Using Angular Routes in a Single-page Application

This tutorial describes how you can build a single-page application that uses multiple Angular routes.

In a single-page application, all of your application's functions exist in a single HTML page.
As users access your application's features, the browser needs to render only the parts that matter to the
user, instead of loading a new page. This pattern can significantly improve your application's
responsiveness.

To define how users navigate through your application, you use routes. You can use routes to
define how users navigate from one part of your application to another. You can also use routes to

> guard against unexpected or unauthorized behavior.

Reviewer 2: I'm all for avoiding the word "use."

   <code-example header="src/app/app.component.html" path="router-tutorial/src/app/app.component.html" region="routeractivelink"></code-example>

View your application again. As you click one of the buttons, the style for that button updates
automatically, identifying the active component to the user. By adding the `routerLinkActive`
directive, you inform your application to apply a specific CSS class to the active route. In this
tutorial, that CSS class is `activebutton`, but you could use any class that you want.

## Adding a redirect

In this step of the tutorial, you add a route that redirects the user to display the `/heroes-list` component.

1. From your code editor, open the `app.module.ts` file.

1. In the `imports` array, update the `RouterModule` section as follows.

   <code-example header="src/app/app.module.ts" path="router-tutorial/src/app/app.module.ts" region="import-redirect"></code-example>

   Notice that this new route uses an empty string as its path. In addition, it replaces the `component` property with two new ones:

   * `redirectTo`. This property instructs Angular to redirect from an empty path to the
     `heroes-list` path.

>    * `pathMatch`. This property instructs Angular to how much of the URL to match. For this

Reviewer 3: ```suggestion
   * `pathMatch`. This property instructs Angular how much of the URL to match. For this
```
Or maybe you meant "on how much"?

Reviewer 2: I meant "on." Thanks!

guard against unexpected or unauthorized behavior.

## Objectives

* Organize a sample application's features into modules.
* Define how to navigate to a component.
* Pass a information to a component using a parameter.
* Structure routes by nesting several routes.
* Check whether users can access a route
* Control whether the application can discard unsave changes.
* Improve performance by pre-fetching route data and lazy loading feature modules.
* Prevent components from loading unless specific criteria are met.

The full source for the final version of the app can be seen and downloaded from the <live-example></live-example>.

## Prerequisites

To complete this tutorial, you should have a basic understanding of the following concepts:

* JavaScript
* HTML
* CSS
* Angular CLI

You might find the Tour of Heroes tutorial helpful, but it is not required.

## Create a sample application

Using the Angular CLI, create a new application, _angular-router-sample_. This application will
have two components: _crisis-list_ and _heroes-list_.

1. Use the `ng new` command to create a new Angular project, _angular-router-sample_.

>    <code-example language="none" class="code-shell">

Reviewer: Why not `language="sh"` (here and below)?

Actually, I think these code examples should be:

```suggestion
   <code-example language="sh">
```

<sub>I know we have the `code-shell` class in many similar code snippets, but it actually adds the same styles as `language="sh"`, so I there shouldn't be a reason not to use that instead.</sub>

   ng new angular-router-sample
   </code-example>

   When prompted with `Would you like to add Angular routing?`, select `N`.

   When prompted with `Which stylesheet format would you like to use?`, select `CSS`.

   After a few moments, a new project, `angular-router-sample`, is ready.

1. From your terminal, navigate to the `angular-router-sample` directory.

> 1. Use the `ng generate` to create a component, _crisis-list_.

Reviewer: ```suggestion
1. Use the `ng generate` command to create a component, _crisis-list_.
```

???

  <code-example language="none" class="code-shell">
   ng generate component crisis-list
  </code-example>

1. In your code editor, locate the file, `crisis-list.component.html` and replace
   the placeholder content with the following HTML.

   <code-example header="src/app/crisis-list/crisis-list.component.html" path="router-tutorial/src/app/crisis-list/crisis-list.component.html"></code-example>

> 1. Use the `ng generate` to create a second component, _heroes-list_.

Reviewer: ```suggestion
1. Use the `ng generate` command to create a second component, _heroes-list_.
```

???

   <code-example header="src/app/app.component.html" path="router-tutorial/src/app/app.component.html" region="setup"></code-example>

1. Verify that your new application runs as expected by running the `ng serve` command.

  <code-example language="none" class="code-shell">
   ng serve
  </code-example>

1. Open a browser to `http://localhost:4200`.

   You should see a single web page, consisting of a title and the HTML of your two components.

## Import `RouterModule` and `Router` from `@angular/router`

Routing allows you to display specific views of your application depending on the URL path. To add this functionality
to your sample application, you need to update the `app.module.ts` file to use two modules: `RouterModule` and
`Router`. You import these modules from `@angular/router`.

1. From your code editor, open the `app.module.ts` file.

1. Add the following `import` statement.

  <code-example header="src/app/app.module.ts" path="router-tutorial/src/app/app.module.ts" region="router-import"></code-example>

## Define your routes

In this section, you'll define two routes:

* The path `[root-path]/crisis-center` opens the `crisis-center` component.

> * The path `[root-path]/heroes-list` opens the `heroes-list` component.

Reviewer: OOC, why the `[root-path]` part?

> A route definition is a JavaScript object with two required propteries. The first property, `path`, is a string

Reviewer: > A route definition is a JavaScript object with two required propteries.

Well, these are not the only properties and they are not required :grin:
(See [here][1].)

[1]: https://angular.io/api/router/Route

## Update your component with `router-outlet`

At this point, you have defined two routes for your application. However, your application
still has both components hard-coded in your `app.component.html` template. For your routes to
work, you need to update your template to dynamically load a component based on the URL path.

To implement this functionality, add the `router-outlet` directive to your template file.

1. From your code editor, open the `app.component.html` file.

1. Delete the following lines.

   <code-example header="src/app/app.component.html" path="router-tutorial/src/app/app.component.html" region="components"></code-example>

1. Add the `router-outlet` directive.

   <code-example header="src/app/app.component.html" path="router-tutorial/src/app/app.component.html" region="router-outlet"></code-example>

View your updated application in your browser. You should see only the application title. To
view the `crisis-list` component, add `crisis-list` to the end of the path in your browser's
address bar. For example:

```
http://localhost:4200/crisis-list

> ```

Reviewer: I think we have decided to stanrdaze to `<code-example>` instead of triple backticks for code blocks.

Notice that the crisis list displays. Angular is using the route you defined to display the
component. You can display the `heroes-list` component the same way:

```
http://localhost:4200/heroes-list
```

## Control navigation with UI elements

Currently, your application supports two routes. However, the only way to use those routes
is for the user to manually type the path in the browser's address bar. In this section, you'll
add two buttons that users can click to navigate between the `heroes-list` and `crisis-list`
components.

> 1. Open the `styles.css` file and add the following styles.

Reviewer: Did you mean `app.component.css`?

1. Open the `app.component.html` file and add the following HTML below the title.

   <code-example header="src/app/app.component.html" path="router-tutorial/src/app/app.component.html" region="nav"></code-example>

   This HTML uses an Angular directive, `routerLink`. This directive connects the routes
   you defined to your template files.

If you view your application in the browser, you should see these two links. When you click
on a link, the corresponding component appears.

## Identify the active route

While users can navigate your application using the links you added in the previous section,
they don't have an easy way to identify what the active route is. You can add this functionality
using Angular's `routerLinkActive` directive.

1. From your code editor, open the `app.component.html` file.

1. Update the anchor tags to include the `routerLinkActive` directive.

   <code-example header="src/app/app.component.html" path="router-tutorial/src/app/app.component.html" region="routeractivelink"></code-example>

View your application again. As you click one of the buttons, the style for that button updates

> automatically, identifying the active component to the user.

Reviewer: I think we should explain how this works. The `routerLinkActive` directive adds the specified class to the `routerLink` that corresponds to the active route. So, the user must define the style for the `.active` element themselves. I think it is worth mentioning that, because otherwise the users might not realize what is going on (and how to adapt ithis to their needs).

## Adding a redirect

> Currently, your `pagenotfound` component displays even when the user first opens the application.

Reviewer: The `pagenotfound` component is introduced in the next section :grin:

Reviewer 2: Rewritten. Thank you!

   Notice that, directly before your wildcard route, you have a new route that uses an empty
   string as its path. In addition, it replaces the `component` property with two new ones:

   * `redirectTo`. This property instructs Angular to redirect from an empty path to the
     `heroes-list` path.
   * `pathMatch`. This property instructs Angular to how much of the URL to match. For this
      tutorial, you should set this property to `full`. This strategy is recommended when
      you have an empty string for a path. For more information about this property,
      see the [Route API documentation](/api/router/Route).

Now when you open your application, it displays the `heroes-list` component by default.

## Adding a 404 page

At present, your application supports two routes. If the user tries to access a page that does not
exist, a best practice is to display a 404 page. In this section, you'll create a 404 page and
update your route configuration to show that page for any unspecified routes.

1. From the terminal, create a new component, `pagenotfound`.

   <code-example language="none" class="code-shell">

>    ng generate component PageNotFound

Reviewer: This command will generate a `PageNotFoundComponent` in `page-not-found.component.*` (instead of `PagenotfoundComponent` in `pagenotfound.component.*`).

<sub>Personally, I prefer what this command generates - I am just pointing out that the command should match the files.</sub>

1. Open the `app.module.ts` file. In the `imports` array, update the `Router` section as follows.

   <code-example header="src/app/app.module.ts" path="router-tutorial/src/app/app.module.ts" region="import-wildcard"></code-example>

   The new route uses a path, `**`. This path is how Angular identifies a wildcard route. Any route
   that does not match an existing route in your configuration will use this route.

   <div class="alert is-important">
    Notice that the wildcard route is placed at the end of the array. The order of your
    routes is important, as Angular applies routes in order and uses the first match it finds.
   </div>

Try navigating to a non-existing route on your application, such as `http://localhost:4200/powers`.
This route doesn't match anything defined in your `app.module.ts` file. However, because you
defined a wildcard route, the application automatically displays your `pagenotfound` component.

<div class="alert is-helpful">
You might also notice that your application displays your <code>pagenotfound</code> component
even when the user first opens the application. This behavior is probably not what you want.

> The next section address this.

Reviewer: ```suggestion
The previous section addresses this.
```

😛

Reviewer 2: Actually, the alert isn't needed any longer, so I removed it.

</div>

## Next steps

At this point, you have a basic application that uses Angular's routing feature to change
what components the user can see based on the URL address. You have extended these features
to account for unexpected or undefined routes, and you've added a wildcard route to display

> a custom 404 page.

Reviewer: > You have extended these features to account for unexpected or undefined routes, and you've added a wildcard route to display a custom 404 page.

Aren't these two the same thing (accounting for undefined routes and adding a wildcard route)?

# Using Angular Routes in a Single-page Application

> This tutorial describes how to build a single-page application that uses multiple Angular routes.

Reviewer 4: Say "describes how you can build..."
```suggestion
This tutorial describes how you can build a single-page application that uses multiple Angular routes.
```

    Notice that the wildcard route is placed at the end of the array. The order of your
    routes is important, as Angular applies routes in order and uses the first match it finds.
   </div>

Try navigating to a non-existing route on your application, such as `http://localhost:4200/powers`.
This route doesn't match anything defined in your `app.module.ts` file. However, because you
defined a wildcard route, the application automatically displays your `pagenotfound` component.

<div class="alert is-helpful">
You might also notice that your application displays your <code>pagenotfound</code> component
even when the user first opens the application. This behavior is probably not what you want.
The next section address this.
</div> 

## Adding a redirect

Currently, your `pagenotfound` component displays even when the user first opens the application.
One way to address this issue is to add a redirect.

1. From your code editor, open the `app.module.ts` file.

1. In the `imports` array, update the `Router` section as follows.

   ```ts
    RouterModule.forRoot([
      {path: 'crisis-list', component: CrisisListComponent},
      {path: 'heroes-list', component: HeroesListComponent},

>       {path: '',    redirectTo: '/heroes-list', pathMatch: 'full' },

Reviewer: ```suggestion
      {path: '', redirectTo: '/heroes-list', pathMatch: 'full'},
```
